### PR TITLE
r/aws_backup_plan: Correctly handle changes to recovery_point_tags

### PR DESCRIPTION
**Pull request by ewbankkit**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Closes https://github.com/terraform-providers/terraform-provider-aws/issues/8431,
Closes https://github.com/terraform-providers/terraform-provider-aws/issues/9502.
Closes https://github.com/terraform-providers/terraform-provider-aws/issues/8193.
Closes https://github.com/terraform-providers/terraform-provider-aws/issues/8737.

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource/aws_backup_plan: Correctly handle changes to recovery_point_tags
```

Output from acceptance testing:

```console
$ make testacc TEST=./aws TESTARGS='-run=TestAccAwsBackupPlan_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAwsBackupPlan_ -timeout 120m
=== RUN   TestAccAwsBackupPlan_basic
=== PAUSE TestAccAwsBackupPlan_basic
=== RUN   TestAccAwsBackupPlan_withTags
=== PAUSE TestAccAwsBackupPlan_withTags
=== RUN   TestAccAwsBackupPlan_withRules
=== PAUSE TestAccAwsBackupPlan_withRules
=== RUN   TestAccAwsBackupPlan_withLifecycle
=== PAUSE TestAccAwsBackupPlan_withLifecycle
=== RUN   TestAccAwsBackupPlan_withRecoveryPointTags
=== PAUSE TestAccAwsBackupPlan_withRecoveryPointTags
=== RUN   TestAccAwsBackupPlan_disappears
=== PAUSE TestAccAwsBackupPlan_disappears
=== CONT  TestAccAwsBackupPlan_basic
=== CONT  TestAccAwsBackupPlan_withRecoveryPointTags
=== CONT  TestAccAwsBackupPlan_withLifecycle
=== CONT  TestAccAwsBackupPlan_disappears
=== CONT  TestAccAwsBackupPlan_withRules
=== CONT  TestAccAwsBackupPlan_withTags
--- PASS: TestAccAwsBackupPlan_disappears (19.47s)
--- PASS: TestAccAwsBackupPlan_basic (21.03s)
--- PASS: TestAccAwsBackupPlan_withRecoveryPointTags (47.45s)
--- PASS: TestAccAwsBackupPlan_withTags (47.99s)
--- PASS: TestAccAwsBackupPlan_withRules (51.20s)
--- PASS: TestAccAwsBackupPlan_withLifecycle (61.37s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	61.428s
```

### CODE REVIEW

File: aws/resource_aws_backup_plan.go [hunk 1]
```diff
@@ -93,152 +90,82 @@ func resourceAwsBackupPlan() *schema.Resource {
...
-		BackupPlan: plan,
-	}
-
-	if v, ok := d.GetOk("tags"); ok {
-		input.BackupPlanTags = tagsFromMapGeneric(v.(map[string]interface{}))
+		BackupPlan: &backup.PlanInput{
+			BackupPlanName: aws.String(d.Get("name").(string)),
+			Rules:          expandBackupPlanRules(d.Get("rule").(*schema.Set)),
+		},
+		BackupPlanTags: keyvaluetags.New(d.Get("tags").(map[string]interface{})).IgnoreAws().BackupTags(),
```
Comment (ewbankkit) [inline aws/resource_aws_backup_plan.go]: Use `keyvaluetags`.

File: aws/resource_aws_backup_plan.go [hunk 2]
```diff
@@ -93,152 +90,82 @@ func resourceAwsBackupPlan() *schema.Resource {
...
 		d.SetId("")
 		return nil
 	}
-
 	if err != nil {
-		return fmt.Errorf("error reading Backup Plan: %s", err)
+		return fmt.Errorf("error reading Backup Plan (%s): %s", d.Id(), err)
 	}
 
-	rule := &schema.Set{F: resourceAwsPlanRuleHash}
```
Comment (ewbankkit) [inline aws/resource_aws_backup_plan.go]: Refactor to `flattenBackupPlanRules()`.

File: aws/resource_aws_backup_plan.go [hunk 3]
```diff
@@ -93,152 +90,82 @@ func resourceAwsBackupPlan() *schema.Resource {
...
+	d.Set("version", resp.VersionId)
 
-		rule.Add(m)
-	}
-	if err := d.Set("rule", rule); err != nil {
+	if err := d.Set("rule", flattenBackupPlanRules(resp.BackupPlan.Rules)); err != nil {
 		return fmt.Errorf("error setting rule: %s", err)
 	}
 
-	tagsOutput, err := conn.ListTags(&backup.ListTagsInput{
```
Comment (ewbankkit) [inline aws/resource_aws_backup_plan.go]: Use `keyvaluetags`.

File: aws/resource_aws_backup_plan.go [hunk 4]
```diff
@@ -93,152 +90,82 @@ func resourceAwsBackupPlan() *schema.Resource {
...
-
-	rules := expandBackupPlanRules(d.Get("rule").(*schema.Set).List())
-
-	plan.Rules = rules
-
-	input := &backup.UpdateBackupPlanInput{
-		BackupPlanId: aws.String(d.Id()),
-		BackupPlan:   plan,
-	}
+	if d.HasChange("rule") {
```
Comment (ewbankkit) [inline aws/resource_aws_backup_plan.go]: Only update if there's actually a change...

File: aws/resource_aws_backup_plan.go [hunk 5]
```diff
@@ -93,152 +90,82 @@ func resourceAwsBackupPlan() *schema.Resource {
...
-		return fmt.Errorf("error updating Backup Plan: %s", err)
+		log.Printf("[DEBUG] Updating Backup Plan: %#v", input)
+		_, err := conn.UpdateBackupPlan(input)
+		if err != nil {
+			return fmt.Errorf("error updating Backup Plan (%s): %s", d.Id(), err)
+		}
 	}
 
 	if d.HasChange("tags") {
-		resourceArn := d.Get("arn").(string)
```
Comment (ewbankkit) [inline aws/resource_aws_backup_plan.go]: Use `keyvaluetags`.

File: aws/resource_aws_backup_plan.go [hunk 6]
```diff
@@ -248,66 +175,64 @@ func resourceAwsBackupPlanUpdate(d *schema.ResourceData, meta interface{}) error
...
 		rule := &backup.RuleInput{}
 
-		if item["rule_name"] != "" {
-			rule.RuleName = aws.String(item["rule_name"].(string))
+		mRule := vRule.(map[string]interface{})
+
+		if vRuleName, ok := mRule["rule_name"].(string); ok && vRuleName != "" {
+			rule.RuleName = aws.String(vRuleName)
+		} else {
+			continue
```
Comment (ewbankkit) [inline aws/resource_aws_backup_plan.go]: This is the same scenario as commented on here: https://github.com/terraform-providers/terraform-provider-aws/pull/6095/files#r226842239.

File: aws/resource_aws_backup_plan.go [hunk 7]
```diff
@@ -316,46 +241,68 @@ func expandBackupPlanRules(l []interface{}) []*backup.RuleInput {
...
+		buf.WriteString(fmt.Sprintf("%d-", v))
+	}
+	if v, ok := mRule["completion_window"].(int); ok {
+		buf.WriteString(fmt.Sprintf("%d-", v))
 	}
 
-	if v, ok := m["start_window"]; ok {
-		buf.WriteString(fmt.Sprintf("%d-", v.(interface{})))
+	if vRecoveryPointTags, ok := mRule["recovery_point_tags"].(map[string]interface{}); ok && len(vRecoveryPointTags) > 0 {
+		buf.WriteString(fmt.Sprintf("%d-", tagsMapToHash(vRecoveryPointTags)))
```
Comment (ewbankkit) [inline aws/resource_aws_backup_plan.go]: Stable has value (not strictly necessary after Go 1.12).

**Comment by bflad** [comment]
This has been released in [version 2.34.0 of the Terraform AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#2340-october-31-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Pull request merged**